lib/active_merchant/billing/gateways/cyber_source_rest.rb

```diff
@@ -129,6 +130,72 @@ def add_address(post, payment_method, address, options, address_type)
         }.compact
       end
 
+      def add_stored_credentials(post, payment_method, options)
+        stored_credential = options[:stored_credential]
+        if stored_credential
+          reason_code = options[:reason_code] || ''
+          options = stored_credential_options(stored_credential, reason_code)
+          post[:commerceIndicator] = options[:transaction_type]
+          stored_credential[:initial_transaction] ? initial_transaction(post, options) : subsequent_transaction(post, options)
+        end
+      end
+
+      def stored_credential_options(options, reason_code)
+        transaction_type = options[:reason_type] || 'internet'
+        initiator = options[:initiator] if  options[:initiator]
+        initiator = 'customer' if initiator == 'cardholder'
+        stored_on_file = options[:reason_type] == 'recurring'
+        options.merge({
+          transaction_type: transaction_type,
+          initiator: initiator,
+          reason_code: reason_code,
+          stored_on_file: stored_on_file
+        })
+      end
+
+      def initial_transaction(post, options)
+        post[:processingInformation] = {
+          authorizationOptions: {
+            initiator: {
+              type: options[:initiator],
+              merchantInitiatedTransaction: {
+                storedCredentialUsed: options[:stored_on_file]
+              }
+            }
+          }
+        }.compact
+      end
+
+      def subsequent_transaction(post, options)
+        processor_information = add_processor_information(post, options)
+        network_transaction_id = processor_information ? processor_information[:transactionId] : ''
+        post[:processingInformation] = {
+          authorizationOptions: {
+            initiator: {
+              type: options[:initiator],
+              merchantInitiatedTransaction: {
+                originalAuthorizedAmount: post.dig(:orderInformation, :amountDetails, :totalAmount), # Discover only
+                previousTransactionID: network_transaction_id,
+                reason: options[:reason_code],
+                storedCredentialUsed: options[:stored_on_file]
+              }
+            }
+          }
+        }.compact
+      end
```

Comment:
I suggest create a new method to add processing information, the base of the hash is the same in two methods. 

e.g.

`def add_processing_information(initiator, merchant_initiated_transaction_hash)
        {
          authorizationOptions: {
            initiator: {
              type: initiator,
              merchantInitiatedTransaction: merchant_initiated_transaction_hash
            }
          }
        }.compact
      end

      def initial_transaction(post, options)
        post[:processingInformation] = add_processing_information(options[:initiator], {
          storedCredentialUsed: options[:stored_on_file]
        })
      end

      def subsequent_transaction(post, options)
        processor_information = add_processor_information(post, options)
        network_transaction_id = processor_information ? processor_information[:transactionId] : ''
        post[:processingInformation] = add_processing_information(options[:initiator], {
          originalAuthorizedAmount: post.dig(:orderInformation, :amountDetails, :totalAmount), # Discover only
          previousTransactionID: network_transaction_id,
          reason: options[:reason_code],
          storedCredentialUsed: options[:stored_on_file]
        })
      end`